File: linkmint-backend/ledger-go/entry.go
package ledger

import (
	"fmt"
	"math/big"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Direction is the side of a ledger leg: DR (debit) or CR (credit).
type Direction string

const (
	DR Direction = "DR"
	CR Direction = "CR"
)

// flip returns the opposite direction (used by Reverse).
func flip(d Direction) Direction {
	if d == CR {
		return DR
	}
	return CR
}

// Leg is one half-entry of a posting: a single account movement. Amount is in minor units
// (NUMERIC(38,0)) and is always strictly positive — the Direction carries the sign.
type Leg struct {
	Account   string
	Direction Direction
	Amount    *big.Int
	Currency  string
}

// PostingInput is a balanced set of legs written atomically under one entry_group. PLID and Note,
// when set, are stamped on every leg of the group (a posting records one business event).
type PostingInput struct {
	EntryGroup *uuid.UUID // optional; a v4 UUID is generated when nil
	Entries    []Leg
	PLID       string
	Note       string
}

// Entry is a persisted ledger row (one leg) as read back from the table.
type Entry struct {
	ID         int64
	EntryGroup uuid.UUID
	Account    string
	Direction  Direction
	Amount     *big.Int
	Currency   string
	PLID       string
	Note       string
	CreatedAt  time.Time
}

// validate enforces the double-entry invariant (A.6) before any write: every leg is well-formed,
// the group has at least one DR and one CR, and DR total == CR total for each currency. A balanced
// multi-currency group is allowed (each currency must balance on its own).
func validate(entries []Leg) error {
	if len(entries) < 2 {
		return fmt.Errorf("%w: a posting needs at least one DR and one CR (got %d legs)", ErrUnbalanced, len(entries))
	}

	type sums struct{ dr, cr *big.Int }
	perCurrency := map[string]*sums{}
	var haveDR, haveCR bool

	for i, leg := range entries {
		if leg.Direction != DR && leg.Direction != CR {
			return fmt.Errorf("%w: leg %d has invalid direction %q", ErrInvalidLeg, i, leg.Direction)
		}
		if strings.TrimSpace(leg.Account) == "" {
			return fmt.Errorf("%w: leg %d has an empty account", ErrInvalidLeg, i)
		}
		if strings.TrimSpace(leg.Currency) == "" {
			return fmt.Errorf("%w: leg %d has an empty currency", ErrInvalidLeg, i)
		}
		if leg.Amount == nil || leg.Amount.Sign() <= 0 {
			return fmt.Errorf("%w: leg %d amount must be a positive integer", ErrInvalidLeg, i)
		}

		s := perCurrency[leg.Currency]
		if s == nil {
			s = &sums{dr: new(big.Int), cr: new(big.Int)}
			perCurrency[leg.Currency] = s
		}
		if leg.Direction == DR {
			s.dr.Add(s.dr, leg.Amount)
			haveDR = true
		} else {
			s.cr.Add(s.cr, leg.Amount)
			haveCR = true
		}
	}

	if !haveDR || !haveCR {
		return fmt.Errorf("%w: a posting needs at least one DR and one CR", ErrUnbalanced)
	}
	for ccy, s := range perCurrency {
		if s.dr.Cmp(s.cr) != 0 {
			return fmt.Errorf("%w: currency %s is unbalanced (DR=%s CR=%s)", ErrUnbalanced, ccy, s.dr, s.cr)
		}
	}
	return nil
}

File: linkmint-backend/ledger-python/src/linkmint_ledger/entry.py
"""Leg/Direction types and the pure balance validator (mirrors ledger-go/entry.go)."""

from __future__ import annotations

import enum
from dataclasses import dataclass

from .errors import InvalidLeg, Unbalanced


class Direction(enum.StrEnum):
    """The side of a ledger leg: DR (debit) or CR (credit)."""

    DR = "DR"
    CR = "CR"


@dataclass(frozen=True)
class Leg:
    """One half-entry of a posting: a single account movement.

    ``amount`` is in minor units (NUMERIC(38,0)) and is always strictly positive — the ``direction``
    carries the sign.
    """

    account: str
    direction: Direction
    amount: int
    currency: str


def _flip(direction: Direction) -> Direction:
    """Return the opposite direction (used by ``reverse``)."""
    return Direction.DR if direction == Direction.CR else Direction.CR


def validate(entries: list[Leg]) -> None:
    """Enforce the double-entry invariant (A.6) before any write.

    Every leg must be well-formed; the group must have at least one DR and one CR; and the DR total
    must equal the CR total for each currency. A balanced multi-currency group is allowed (each
    currency must balance on its own). Raises :class:`Unbalanced` or :class:`InvalidLeg`.
    """
    if len(entries) < 2:
        raise Unbalanced(f"a posting needs at least one DR and one CR (got {len(entries)} legs)")

    per_currency: dict[str, list[int]] = {}  # currency -> [dr_total, cr_total]
    have_dr = have_cr = False

    for i, leg in enumerate(entries):
        try:
            direction = Direction(leg.direction)
        except ValueError:
            raise InvalidLeg(f"leg {i} has invalid direction {leg.direction!r}") from None
        if not leg.account or not leg.account.strip():
            raise InvalidLeg(f"leg {i} has an empty account")
        if not leg.currency or not leg.currency.strip():
            raise InvalidLeg(f"leg {i} has an empty currency")
        # bool is a subclass of int — reject it explicitly so True/False can't be an amount.
        if isinstance(leg.amount, bool) or not isinstance(leg.amount, int) or leg.amount <= 0:
            raise InvalidLeg(f"leg {i} amount must be a positive integer")

        sums = per_currency.setdefault(leg.currency, [0, 0])
        if direction == Direction.DR:
            sums[0] += leg.amount
            have_dr = True
        else:
            sums[1] += leg.amount
            have_cr = True

    if not have_dr or not have_cr:
        raise Unbalanced("a posting needs at least one DR and one CR")
    for ccy, (dr, cr) in per_currency.items():
        if dr != cr:
            raise Unbalanced(f"currency {ccy} is unbalanced (DR={dr} CR={cr})")
